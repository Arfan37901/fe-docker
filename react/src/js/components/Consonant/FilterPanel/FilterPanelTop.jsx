import React from 'react';
import PropTypes from 'prop-types';
import FilterItem from '../FilterItem';

const DESKTOP_MIN_WIDTH = 1200;
const TABLET_MIN_WIDTH = 768;
const FiltersPanelTop = (props) => {
    const {
        filters,
        resQty,
        showTotalResults,
        showSearchbar,
        onCheckboxClick,
        onFilterClick,
        onClearAllFilters,
        onClearFilterItems,
        clearFilterText,
        clearAllFiltersText,
        children,
        showLimitedFiltersQty,
        onShowAllClick,
    } = props;

    let updatedChildren = [];
    const renderChildren = (key) => {
        const res = updatedChildren.filter(el => el.props && el.props.childrenKey === key);
        return res.length > 0 ? res : null;
    };
    const countSelectedInFilter = el => el.reduce((acc, val) => (val.selected ? acc + 1 : acc), 0);
    const checkFiltersSelected = () => filters.some(f => countSelectedInFilter(f.items) > 0);

    if (!Array.isArray(children)) updatedChildren.push(children);
    else updatedChildren = children;

    return (
        <div data-testid="consonant-filters__top" className="consonant-top-filters">
            {
                updatedChildren.some(el => el.props && el.props.childrenKey === 'filtersTopSearch') &&
                window.innerWidth < TABLET_MIN_WIDTH &&
                    <div data-testid="top-filters__search-wrapper" className="consonant-top-filters--search-wrapper">
                        {renderChildren('filtersTopSearch')}
                    </div>
            }
            <div className="consonant-top-filters--inner">
                {filters.length > 0 &&
                    <div className="consonant-top-filters--filters-wrapper">
                        {window.innerWidth >= TABLET_MIN_WIDTH &&
                            <strong className="consonant-top-filters--title">Filters:</strong>
                        }
                        <div
                            data-testid="consonant-filters__top__filters"
                            className={
                                showLimitedFiltersQty ?
                                    'consonant-top-filters--filters consonant-top-filters--filters_truncated' :
                                    'consonant-top-filters--filters'
                            }>
                            {filters.map(item =>
                                (<FilterItem
                                    key={item.id}
                                    name={item.group}
                                    icon={item.icon}
                                    items={item.items}
                                    itemsSelected={countSelectedInFilter(item.items)}
                                    results={resQty}
                                    id={item.id}
                                    isOpened={item.opened}
                                    onCheck={onCheckboxClick}
                                    onClick={onFilterClick}
                                    onClearAll={onClearFilterItems}
                                    clearFilterText={clearFilterText}
                                    isTopFilter />))
                            }
                            {
                                filters.length > 2 &&
                                window.innerWidth < DESKTOP_MIN_WIDTH &&
                                window.innerWidth >= TABLET_MIN_WIDTH &&
                                <button
                                    type="button"
                                    data-testid="top-filter__more-button"
                                    className="consonant-top-filters--more-btn"
                                    onClick={onShowAllClick}>
                                    {showLimitedFiltersQty ? 'more filters +' : 'hide -'}
                                </button>
                            }
                        </div>
                        <div className="consonant-top-filters--clear-btn-wrapper">
                            {checkFiltersSelected() &&
                            <button
                                type="button"
                                data-testid="top-filter__clear-button"
                                className="consonant-top-filters--clear-btn"
                                onClick={onClearAllFilters}>{clearAllFiltersText}
                            </button>
                            }
                        </div>
                    </div>
                }
                {window.innerWidth >= TABLET_MIN_WIDTH && showTotalResults &&
                    <span
                        data-testid="filter-top-result-count"
                        className="consonant-top-filters--res-qty">
                        <strong>{resQty} </strong>results
                    </span>
                }
                {
                    updatedChildren.some(el => el.props && el.props.childrenKey === 'filtersTopSearchIco') &&
                    window.innerWidth >= TABLET_MIN_WIDTH &&
                        <div
                            data-testid="filter-top-ico-wrapper"
                            className="consonant-top-filters--search-ico-wrapper">
                            {
                                showSearchbar ?
                                    renderChildren('filtersTopSearch') :
                                    renderChildren('filtersTopSearchIco')
                            }
                        </div>
                }
                {updatedChildren.some(el => el.props && el.props.childrenKey === 'filtersTopSelect') &&
                    <div data-testid="top-filters__select-wrapper" className="consonant-top-filters--select-wrapper">
                        {renderChildren('filtersTopSelect')}
                    </div>
                }
            </div>
        </div>
    );
};

export default FiltersPanelTop;

FiltersPanelTop.propTypes = {
    children: PropTypes.oneOfType([
        PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.bool, PropTypes.element])),
        PropTypes.element,
        PropTypes.bool,
    ]),
    filters: PropTypes.arrayOf(PropTypes.object),
    resQty: PropTypes.number,
    onCheckboxClick: PropTypes.func.isRequired,
    onFilterClick: PropTypes.func.isRequired,
    onClearAllFilters: PropTypes.func.isRequired,
    onClearFilterItems: PropTypes.func.isRequired,
    clearFilterText: PropTypes.string.isRequired,
    clearAllFiltersText: PropTypes.string.isRequired,
    showTotalResults: PropTypes.bool,
    showSearchbar: PropTypes.bool,
    showLimitedFiltersQty: PropTypes.bool,
    onShowAllClick: PropTypes.func.isRequired,
};

FiltersPanelTop.defaultProps = {
    filters: [],
    resQty: 0,
    showTotalResults: true,
    showSearchbar: false,
    children: [],
    showLimitedFiltersQty: false,
};
