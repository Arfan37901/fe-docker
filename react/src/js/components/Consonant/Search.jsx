import React from 'react';
import PropTypes from 'prop-types';

const Search = (props) => {
    const { itemsQty, value, onSearch } = props;
    const handleSearch = (evt) => {
        onSearch(evt.target.value);
    };

    const clearSearch = () => {
        onSearch('');
    };

    return (
        <div className="consonant-search">
            <label>
                <span className="consonant-search--input-title">Search</span>
                <span className="consonant-search--input-wrapper">
                    <input
                        type="search"
                        placeholder={onSearch ? `Search within ${itemsQty} items` : 'Search'}
                        value={value}
                        onChange={handleSearch}
                        required />
                    <button
                        type="button"
                        title="Click to clear search query"
                        className="consonant-search--input-clear"
                        onClick={clearSearch}>clear
                    </button>
                </span>
            </label>
        </div>
    );
};

export default Search;

Search.propTypes = {
    itemsQty: PropTypes.number,
    onSearch: PropTypes.func.isRequired,
    value: PropTypes.string.isRequired,
};

Search.defaultProps = {
    itemsQty: 0,
};
