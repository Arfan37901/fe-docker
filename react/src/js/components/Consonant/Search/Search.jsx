import PropTypes from 'prop-types';
import React from 'react';

const searchId = 'consonant-search';
const Search = (props) => {
    const {
        value,
        onSearch,
        placeholderText,
        leftPanelTitle,
        name,
        autofocus,
    } = props;

    let textInput = null;
    const handleSearch = (evt) => {
        onSearch(evt.target.value);
    };
    const setAutofocus = () => {
        textInput.focus();
    };
    const clearSearch = () => {
        onSearch('');
        setAutofocus();
    };
    return (
        <div data-testid={name} className="consonant-search">
            <label htmlFor={searchId}>
                <span className="consonant-search--input-title">{leftPanelTitle}</span>
                <span className="consonant-search--input-wrapper">
                    <input
                        id={searchId}
                        data-testid="search-input"
                        type="search"
                        placeholder={placeholderText}
                        onClick={e => e.stopPropagation()}
                        value={value}
                        onChange={handleSearch}
                        ref={(input) => {
                            if (autofocus && input) input.focus();
                            textInput = input;
                        }}
                        className="consonant-search--input"
                        required />
                    <button
                        type="button"
                        title="Click to clear search query"
                        className="consonant-search--input-clear"
                        onClick={clearSearch}
                        tabIndex="0">clear
                    </button>
                </span>
            </label>
        </div>
    );
};

export default Search;

Search.propTypes = {
    name: PropTypes.string,
    onSearch: PropTypes.func.isRequired,
    value: PropTypes.string,
    placeholderText: PropTypes.string,
    leftPanelTitle: PropTypes.string,
    autofocus: PropTypes.bool,
};

Search.defaultProps = {
    value: '',
    name: 'consonant-search',
    placeholderText: 'Search here...',
    leftPanelTitle: 'Search',
    autofocus: true,
};
