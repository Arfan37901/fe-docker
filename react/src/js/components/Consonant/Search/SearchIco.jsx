import React from 'react';
import PropTypes from 'prop-types';

const SearchIco = (props) => {
    const { onClick } = props;

    return (
        <button
            type="button"
            className="search-ico"
            onClick={onClick}
            tabIndex="0">
            Click to search
            <span />
        </button>
    );
};

export default SearchIco;

SearchIco.propTypes = {
    onClick: PropTypes.func.isRequired,
};
