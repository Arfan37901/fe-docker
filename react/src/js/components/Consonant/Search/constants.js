export const SearchInputDefaultProps = {
    name: '',
    value: '',
    autofocus: true,
    placeholderText: '',
};

/**
 * Used as unique id for accessibility labels/attributes
 * @type {Number}
 */
export const searchInputId = 'consonant-search';
