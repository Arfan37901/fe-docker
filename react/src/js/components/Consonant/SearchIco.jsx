import React from 'react';
import PropTypes from 'prop-types';

const SearchIco = (props) => {
    const { onClick } = props;

    return (
        <button
            type="button"
            className="search-ico"
            onClick={onClick}>
                Click to search<span />
        </button>
    );
};

export default SearchIco;

SearchIco.propTypes = {
    onClick: PropTypes.func.isRequired,
};
