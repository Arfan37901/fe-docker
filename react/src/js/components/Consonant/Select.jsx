import React from 'react';
import PropTypes from 'prop-types';

const Select = (props) => {
    const { val, values, onSelect } = props;

    return (
        <div className="consonant-select">
            <input
                type="text"
                value={val.label || 'Please select'}
                readOnly />
            <div className="consonant-select--options">
                {values.map(item => (
                    <button
                        key={item.label}
                        type="button"
                        className={item.label === val.label ?
                            'consonant-select--option consonant-select--option_selected' :
                            'consonant-select--option'
                        }
                        onClick={() => { onSelect(item); }}>
                        {item.label}
                    </button>
                ))}
            </div>
        </div>
    );
};

export default Select;

Select.propTypes = {
    val: PropTypes.shape({
        label: PropTypes.string,
        sort: PropTypes.string,
    }).isRequired,
    onSelect: PropTypes.func.isRequired,
    values: PropTypes.arrayOf(PropTypes.object).isRequired,
};
