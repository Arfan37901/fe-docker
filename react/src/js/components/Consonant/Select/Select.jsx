import React from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';
import { useExpandable } from '../../../utils/hooks';

const Select = ({
    val,
    values,
    onSelect,
    autoWidth,
    optionsAlignment,
    id,
}) => {
    const [openDropdown, handleToggle] = useExpandable(id);
    const opened = openDropdown === id;

    const handleOptionClick = (e, item) => {
        onSelect(item);
        handleToggle(e);
    };

    const shouldAutoWidthSortClass = classNames({
        'consonant-select': true,
        'consonant-select_auto-width': autoWidth,
    });

    const openButtonClass = classNames({
        'consonant-select--btn': true,
        'consonant-select--btn_active': opened,
    });

    return (
        <div className={shouldAutoWidthSortClass}>
            <button
                data-testid="select-button"
                type="button"
                onClick={handleToggle}
                className={openButtonClass}
                tabIndex="0">
                {val.label}
            </button>
            <div
                data-testid="consonant-select--options"
                className={`consonant-select--options consonant-select--options_${optionsAlignment}`}>
                {values.map(item => (
                    <button
                        data-testid="select-option"
                        key={item.label}
                        type="button"
                        className={item.label === val.label ?
                            'consonant-select--option consonant-select--option_selected' :
                            'consonant-select--option'
                        }
                        onClick={e => handleOptionClick(e, item)}
                        tabIndex={opened ? '0' : '-1'}>
                        {item.label}
                    </button>
                ))}
            </div>
        </div>
    );
};

export default Select;

Select.propTypes = {
    val: PropTypes.shape({
        label: PropTypes.string,
        sort: PropTypes.string,
    }).isRequired,
    onSelect: PropTypes.func.isRequired,
    values: PropTypes.arrayOf(PropTypes.object).isRequired,
    autoWidth: PropTypes.bool,
    optionsAlignment: PropTypes.string,
    id: PropTypes.string.isRequired,
};

Select.defaultProps = {
    autoWidth: false,
    optionsAlignment: 'right',
};
