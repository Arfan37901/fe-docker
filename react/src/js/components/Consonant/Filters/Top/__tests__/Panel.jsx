import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import FilterPanelTop from '../Panel';
import Select from '../../../Select/Select';
import Search from '../../../Search/Search';
import SearchIco from '../../../Search/SearchIco';

import {
    DEFAULT_PROPS,
    TABLET_MIN_WIDTH,
    selectedAllFilters,
    MOBILE_MIN_WIDTH,
} from '../../../Helpers/Testing/Constants/FilterPanelTop';

import { DEFAULT_PROPS as SEARCH_DEFAULT_PROPS } from '../../../Helpers/Testing/Constants/Search';
import { DEFAULT_PROPS as SELECT_DEFAULT_PROPS } from '../../../Helpers/Testing/Constants/Select';

import makeSetup, { createTree } from '../../../Helpers/Testing/Utils/Settings';

const setup = makeSetup(FilterPanelTop, DEFAULT_PROPS);

const multipleFilters = [...DEFAULT_PROPS.filters, ...DEFAULT_PROPS.filters]
    .map((item, index) => ({ ...item, id: `${item}_${index}` }));

describe('Top Filter Panel', () => {
    beforeEach(() => {
        global.innerWidth = TABLET_MIN_WIDTH;
    });

    test('Should render all groups of filters', () => {
        const { props: { filters } } = setup();

        const filterGroupElements = screen.queryAllByTestId('filter-group');

        expect(filterGroupElements).toHaveLength(filters.length);
    });
    test('Should display total results if authored', () => {
        setup({ displayTotalResults: true });

        const footerTotalResElement = screen.queryByTestId('filter-top-result-count');

        expect(footerTotalResElement).not.toBeNull();
    });

    test('Should be able to render a search box on mobile', () => {
        global.innerWidth = MOBILE_MIN_WIDTH;

        render((
            <FilterPanelTop
                {...DEFAULT_PROPS}
                searchComponent={(
                    <Search
                        name="filtersTopSearch"
                        {...SEARCH_DEFAULT_PROPS} />
                )} />
        ));

        const footerTotalResElement = screen.queryByTestId('top-filters__search-wrapper');

        expect(footerTotalResElement).not.toBeNull();
    });
    test('Should renders correctly without search', () => {
        render((
            <FilterPanelTop {...DEFAULT_PROPS} >
                <Search
                    name="filtersTopSearch"
                    {...SEARCH_DEFAULT_PROPS} />
            </FilterPanelTop>
        ));

        const footerTotalResElement = screen.queryByTestId('top-filters__search-wrapper');

        expect(footerTotalResElement).toBeNull();
    });
    test('Should be able to show the Sort Popup', () => {
        render((
            <FilterPanelTop
                {...DEFAULT_PROPS}
                sortEnabled
                sortOptions={[{ label: 'Featured', sort: 'featured' }]}
                sortComponent={(
                    <Select
                        childrenKey="filtersTopSelect"
                        {...SELECT_DEFAULT_PROPS} />
                )} />
        ));

        const footerTotalResElement = screen.queryByTestId('top-filters__sort-popup');

        expect(footerTotalResElement).not.toBeNull();
    });

    test('should renders more button with `more filters +` text', () => {
        setup({ showLimitedFiltersQty: true, filters: multipleFilters });

        const footerTotalResElement = screen.queryByTestId('top-filter__more-button');

        expect(footerTotalResElement).toHaveTextContent('more filters +');
    });

    test('Filter Group Button Should Exist', () => {
        setup();
        const [filterGroupBtn] = screen.queryAllByTestId('filter-group-btn');
        expect(filterGroupBtn).toBeDefined();
    });

    test('should show clear all filters text', () => {
        setup({ filters: selectedAllFilters });

        const clearButtonElement = screen.queryByTestId('top-filter__clear-button');

        expect(clearButtonElement).not.toBeNull();
    });

    describe('Check snapshots', () => {
        test('The Top Filter Should Show A Search Icon', () => {
            const tree = createTree((
                <FilterPanelTop {...DEFAULT_PROPS} >
                    <SearchIco
                        childrenKey="filtersTopSearchIco" />
                </FilterPanelTop>
            ));

            expect(tree).toMatchSnapshot();
        });
        test('should renders correctly without children', () => {
            const tree = createTree((
                <FilterPanelTop {...DEFAULT_PROPS} />
            ));

            expect(tree).toMatchSnapshot();
        });
        test('Shoulld renders correctly on desktop', () => {
            const { tree } = setup();

            expect(tree).toMatchSnapshot();
        });
        test('should renders correctly with limited filters', () => {
            const { tree } = setup({ showLimitedFiltersQty: 10 });

            expect(tree).toMatchSnapshot();
        });
        test('should renders correctly with more button', () => {
            const { tree } = setup({ filters: multipleFilters });

            expect(tree).toMatchSnapshot();
        });
        test('should renders correctly with non-table width', () => {
            global.innerWidth = MOBILE_MIN_WIDTH;

            const { tree } = setup();

            expect(tree).toMatchSnapshot();
        });
    });

    describe('Testing UI Interactions', () => {
        test('When "Clear All Filters" is clicked, the appropriate event handler should be called', () => {
            const {
                props: {
                    onClearAllFilters,
                },
            } = setup({ filters: selectedAllFilters });

            const clearButtonElement = screen.queryByTestId('top-filter__clear-button');

            fireEvent.click(clearButtonElement);

            expect(onClearAllFilters).toBeCalled();
        });
    });
});
