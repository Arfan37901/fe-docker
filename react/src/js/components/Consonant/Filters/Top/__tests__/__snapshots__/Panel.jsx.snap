// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Top Filter Panel Check snapshots Shoulld renders correctly on desktop 1`] = `
<div
  className="consonant-top-filters"
  data-testid="consonant-filters__top"
>
  <div
    className="consonant-top-filters--inner"
  >
    <span
      className="consonant-top-filters--res-qty"
      data-testid="filter-top-result-count"
    >
      <strong>
        0 results
      </strong>
    </span>
    <div
      className="consonant-top-filters--search-ico-wrapper"
      data-testid="filter-top-ico-wrapper"
    >
      <button
        className="search-ico"
        onClick={[Function]}
        tabIndex="0"
        type="button"
      >
        Click to search
        <span />
      </button>
    </div>
  </div>
</div>
`;

exports[`Top Filter Panel Check snapshots The Top Filter Should Show A Search Icon 1`] = `
<div
  className="consonant-top-filters"
  data-testid="consonant-filters__top"
>
  <div
    className="consonant-top-filters--inner"
  >
    <span
      className="consonant-top-filters--res-qty"
      data-testid="filter-top-result-count"
    >
      <strong>
        0 results
      </strong>
    </span>
    <div
      className="consonant-top-filters--search-ico-wrapper"
      data-testid="filter-top-ico-wrapper"
    >
      <button
        className="search-ico"
        onClick={[Function]}
        tabIndex="0"
        type="button"
      >
        Click to search
        <span />
      </button>
    </div>
  </div>
</div>
`;

exports[`Top Filter Panel Check snapshots should renders correctly with limited filters 1`] = `
<div
  className="consonant-top-filters"
  data-testid="consonant-filters__top"
>
  <div
    className="consonant-top-filters--inner"
  >
    <span
      className="consonant-top-filters--res-qty"
      data-testid="filter-top-result-count"
    >
      <strong>
        0 results
      </strong>
    </span>
    <div
      className="consonant-top-filters--search-ico-wrapper"
      data-testid="filter-top-ico-wrapper"
    >
      <button
        className="search-ico"
        onClick={[Function]}
        tabIndex="0"
        type="button"
      >
        Click to search
        <span />
      </button>
    </div>
  </div>
</div>
`;

exports[`Top Filter Panel Check snapshots should renders correctly with more button 1`] = `
<div
  className="consonant-top-filters"
  data-testid="consonant-filters__top"
>
  <div
    className="consonant-top-filters--inner"
  >
    <span
      className="consonant-top-filters--res-qty"
      data-testid="filter-top-result-count"
    >
      <strong>
        0 results
      </strong>
    </span>
    <div
      className="consonant-top-filters--search-ico-wrapper"
      data-testid="filter-top-ico-wrapper"
    >
      <button
        className="search-ico"
        onClick={[Function]}
        tabIndex="0"
        type="button"
      >
        Click to search
        <span />
      </button>
    </div>
  </div>
</div>
`;

exports[`Top Filter Panel Check snapshots should renders correctly with non-table width 1`] = `
<div
  className="consonant-top-filters"
  data-testid="consonant-filters__top"
>
  <div
    className="consonant-top-filters--inner"
  >
    <span
      className="consonant-top-filters--res-qty"
      data-testid="filter-top-result-count"
    >
      <strong>
        0 results
      </strong>
    </span>
    <div
      className="consonant-top-filters--search-ico-wrapper"
      data-testid="filter-top-ico-wrapper"
    >
      <button
        className="search-ico"
        onClick={[Function]}
        tabIndex="0"
        type="button"
      >
        Click to search
        <span />
      </button>
    </div>
  </div>
</div>
`;

exports[`Top Filter Panel Check snapshots should renders correctly without children 1`] = `
<div
  className="consonant-top-filters"
  data-testid="consonant-filters__top"
>
  <div
    className="consonant-top-filters--inner"
  >
    <span
      className="consonant-top-filters--res-qty"
      data-testid="filter-top-result-count"
    >
      <strong>
        0 results
      </strong>
    </span>
    <div
      className="consonant-top-filters--search-ico-wrapper"
      data-testid="filter-top-ico-wrapper"
    >
      <button
        className="search-ico"
        onClick={[Function]}
        tabIndex="0"
        type="button"
      >
        Click to search
        <span />
      </button>
    </div>
  </div>
</div>
`;
